Improve Mark formatting

geriux opened the pull request:
**Related PRs:**
- https://github.com/wordpress-mobile/gutenberg-mobile/pull/6517
- https://github.com/WordPress/gutenberg/pull/57650
- https://github.com/wordpress-mobile/WordPress-iOS/pull/22425

This PR fixes several issues related to the Mark formatting.

- It completes the missing functionality implementation to fully support the Mark tag.
- Adds `MarkStringAttributeConverter` which will handle toggling the format as well as returning the expected HTML format to Gutenberg, by removing the extra spaces. This was causing issues like history changes or showing the active style indicator when it shouldn't.
- Updates `MarkFormatter` to set the expected CSS attributes as well as adding a `color` param.  
- Adds `preprocessMarkForInsertion` that will handle cases like auto-corrected strings, it will check if the new auto-corrected words need to have the Mark representation along with `adjustAttributesForMark`.
- It also fixes `toggleMark` by allowing passing a `color` param and to be able to change colors while editing a Mark formatting or resetting the existing format.

## To test

Check this [PR description](https://github.com/WordPress/gutenberg/pull/57650) under the `Test` section. Where there are builds linked as well.

Commit:
Mark Formatting - Fix autocorrected strings and changing the caret position bug

Commit:
Mark formatter - Fix autocorrected formatting

twstokes commented:
> Hey @twstokes 👋 I was wondering if you could help me out with a review for this PR since part of it uses the same code you added for the Heading fix for the highlighting text feature.
> 
> No worries if you can't, let me know either way. Thanks! 🙇

Hey @geriux! 👋  I'll be glad to. 👍

geriux commented:
> > Hey @twstokes 👋 I was wondering if you could help me out with a review for this PR since part of it uses the same code you added for the Heading fix for the highlighting text feature.
> > No worries if you can't, let me know either way. Thanks! 🙇
> 
> Hey @geriux! 👋 I'll be glad to. 👍

🎉 thank you!

twstokes commented on Aztec/Classes/TextKit/TextStorage.swift:
```diff
@@ -253,6 +254,46 @@ open class TextStorage: NSTextStorage {
+    ///
+    /// - Parameters:
+    ///     - attributedString: the string we need to preprocess.
+    ///
```
Nit: Missing param for `_ range: NSRange`.

twstokes commented on Aztec/Classes/TextKit/TextStorage.swift:
```diff
@@ -253,6 +254,46 @@ open class TextStorage: NSTextStorage {
+    /// - Returns: the preprocessed string.
+    ///
+    fileprivate func preprocessMarkForInsertion(_ attributedString: NSAttributedString, _ range: NSRange) -> NSAttributedString {
+        guard textStore.length > 0, attributedString.length > 0, range.endLocation != range.lowerBound else {
```
Is `range.endLocation != range.lowerBound` equivalent to `range.length > 0`?

`range.endLocation` is the same as `range.location` + `range.length`, and I believe `range.lowerBound` is equivalent to `range.location`. So it seems they can only be equivalent if length is zero, but I may be missing something!

twstokes commented on Aztec/Classes/TextKit/TextStorage.swift:
```diff
@@ -253,6 +254,46 @@ open class TextStorage: NSTextStorage {
+        }
+        var startAt = 0
+
+        if range.endLocation > range.lowerBound {
```
I may need some help on this one as well - is there an instance where the end location won't be higher than the lower bound for lengths > 0?

twstokes commented on Aztec/Classes/Formatters/Implementations/MarkFormatter.swift:
```diff
@@ -24,8 +24,8 @@ class MarkFormatter: AttributeFormatter {
     func remove(from attributes: [NSAttributedString.Key: Any]) -> [NSAttributedString.Key: Any] {
         var resultingAttributes = attributes
 
+        resultingAttributes[.foregroundColor] = placeholderAttributes![.foregroundColor]
```
Are we sure `placeholderAttributes` is safe to force unwrap every time or should we do the `guard let` pattern?

twstokes reviewed (commented):
👋 @geriux! I had a few questions in the PR, but otherwise the tests from the [Gutenberg PR](https://github.com/WordPress/gutenberg/pull/39454) passed. 👍 

I did notice a few things that looked like bugs that I wanted to share. They may not be due to this PR and you may already be aware of them. Unfortunately, only two were reproducible.

1. When viewing the HTML output, I found that one of my `<mark>` elements had two `color` properties. [couldn't reproduce]

<img src="https://user-images.githubusercontent.com/2092798/159201343-b80438c2-ec3e-4998-8e84-3823d2f9e140.PNG" width="350px" alt="Two color properties" />

2. The formatting bar retained the highlighted color even when it wasn't being applied. [reproducible]

https://user-images.githubusercontent.com/2092798/159201417-323294a1-d06b-471a-9752-92168eff178d.MP4

3. Starting with an empty paragraph block, choosing a highlight color, and typing resulted in no highlighted text. It was as if I'd never chosen the color. I didn't get a chance to grab a video of this one. [couldn't reproduce]

4. Text color didn't match the custom color chosen. I confirmed this also happens in the latest TestFlight version 19.4.0.3. This was a simple site running the Blank Canvas theme. It only happened with custom colors chosen at the bottom of the picker, not the custom color picker in the "Theme" section. [reproducible]

https://user-images.githubusercontent.com/2092798/159201625-ea512f9b-2453-4242-9a63-677d5fed831a.MP4

geriux commented:
> 👋 Hey @geriux! Is there anything I can do to help with this PR or have we decided not to pursue it for now? Thanks!

Hey @twstokes Thank you for reaching out! Due to other priorities, I couldn't focus on addressing your feedback but I will go back to this so we can include this fix. 🙇

twstokes commented:
> > 👋 Hey @geriux! Is there anything I can do to help with this PR or have we decided not to pursue it for now? Thanks!
> 
> Hey @twstokes Thank you for reaching out! Due to other priorities, I couldn't focus on addressing your feedback but I will go back to this so we can include this fix. 🙇

Sounds good!

Commit:
Merge branch 'develop' into feature/mark-formatting-improvements

Commit:
Remove setting the color to the placeholder attributes

Commit:
Update preprocessMarkForInsertion

Commit:
Fix whitespace issues

Commit:
Update MarkFormatter to support incoming color changes

Commit:
Update how toggleMark works to support color customization and resetting the format

Commit:
Update TextStorage to process the mark formatting within replaceCharacters

Commit:
Fix Hound lint issues

Commit:
Fix lint issue

Commit:
Update MarkConverter to remove extra spaces within the style attribute

geriux commented:
Hey there @twstokes 👋 I'm sorry this took so long to move forward but I've introduced new changes that should fix the issues you mentioned above 🚀

twstokes commented on Aztec/Classes/Formatters/Implementations/MarkFormatter.swift:
```diff
@@ -4,28 +4,43 @@ import UIKit
         }
         resultingAttributes[.markHtmlRepresentation] = representationToUse
 
+        if textColor != nil {
```
Optional nit: I'd consider an `if let` here to unwrap `textColor` safely so we don't have to force unwrap `textColor` later. It should be safe as is, just a preferred idiom.

geriux replied:
Thanks for the suggestion! Addressed in https://github.com/wordpress-mobile/AztecEditor-iOS/pull/1352/commits/28e62ab8f254eec55b87e4a5607e2c9d4a9887d4

twstokes commented on Aztec/Classes/Formatters/Implementations/MarkFormatter.swift:
```diff
@@ -4,28 +4,43 @@ import UIKit
 
     var placeholderAttributes: [NSAttributedString.Key: Any]?
+    var textColor: String?
+    var defaulTextColor: UIColor?
```
Optional nit:
```suggestion
    var defaultTextColor: UIColor?
```

geriux replied:
Nice catch! Updated in https://github.com/wordpress-mobile/AztecEditor-iOS/pull/1352/commits/7b9cc634d514835a0fd92d7743123211b06da4ee

twstokes reviewed (approved):
Great work @geriux! 🚀 This makes sense to me, and I added two comments that are optional nits. 

I ran through a couple tests, but mostly relied on David and Carlos' results [in this PR](https://github.com/WordPress/gutenberg/pull/57650). If you think I should do all of the manual tests as well I'd be happy to.

Commit:
Use if let to unwrap textColor value

Commit:
Fix defaultTextColor typo

twstokes reviewed (approved):
Looks great! 🙇
